packages/session-replay-browser/src/session-replay.ts

```diff
@@ -0,0 +1,530 @@
+import { getAnalyticsConnector, getGlobalScope } from '@amplitude/analytics-client-common';
+import { BaseTransport, Logger } from '@amplitude/analytics-core';
+import { Logger as ILogger, ServerZone, Status } from '@amplitude/analytics-types';
+import * as IDBKeyVal from 'idb-keyval';
+import { pack, record } from 'rrweb';
+import { SessionReplayConfig } from './config';
+import {
+  BLOCK_CLASS,
+  DEFAULT_SESSION_REPLAY_PROPERTY,
+  MASK_TEXT_CLASS,
+  MAX_EVENT_LIST_SIZE_IN_BYTES,
+  MAX_IDB_STORAGE_LENGTH,
+  MAX_INTERVAL,
+  MIN_INTERVAL,
+  SESSION_REPLAY_EU_URL as SESSION_REPLAY_EU_SERVER_URL,
+  SESSION_REPLAY_SERVER_URL,
+  STORAGE_PREFIX,
+  defaultSessionStore,
+} from './constants';
+import { isSessionInSample, maskInputFn } from './helpers';
+import {
+  MAX_RETRIES_EXCEEDED_MESSAGE,
+  MISSING_API_KEY_MESSAGE,
+  MISSING_DEVICE_ID_MESSAGE,
+  STORAGE_FAILURE,
+  UNEXPECTED_ERROR_MESSAGE,
+  getSuccessMessage,
+} from './messages';
+import {
+  AmplitudeSessionReplay,
+  Events,
+  IDBStore,
+  IDBStoreSession,
+  SessionReplayConfig as ISessionReplayConfig,
+  RecordingStatus,
+  SessionReplayContext,
+  SessionReplayOptions,
+} from './typings/session-replay';
+
+export class SessionReplay implements AmplitudeSessionReplay {
+  name = '@amplitude/session-replay-browser';
+  config: ISessionReplayConfig | undefined;
+  loggerProvider: ILogger;
+  storageKey = '';
+  retryTimeout = 1000;
+  events: Events = [];
+  currentSequenceId = 0;
+  private scheduled: ReturnType<typeof setTimeout> | null = null;
+  queue: SessionReplayContext[] = [];
+  stopRecordingEvents: ReturnType<typeof record> | null = null;
+  maxPersistedEventsSize = MAX_EVENT_LIST_SIZE_IN_BYTES;
+  interval = MIN_INTERVAL;
+  timeAtLastSend: number | null = null;
+
+  constructor() {
+    this.loggerProvider = new Logger();
+  }
+
+  async init(apiKey: string, options: SessionReplayOptions) {
+    this.config = new SessionReplayConfig(apiKey, options);
+    this.loggerProvider = this.config.loggerProvider;
+
+    this.loggerProvider.log('Installing @amplitude/session-replay-browser.');
+
+    this.storageKey = `${STORAGE_PREFIX}_${this.config.apiKey.substring(0, 10)}`;
+
+    const globalScope = getGlobalScope();
+    if (globalScope) {
+      globalScope.addEventListener('blur', this.blurListener);
+      globalScope.addEventListener('focus', this.focusListener);
+    }
+
+    if (globalScope && globalScope.document && globalScope.document.hasFocus()) {
+      await this.initialize(true);
+    }
+  }
+
+  setSessionId(sessionId: number) {
```

Comment:
hmm we have setSessionId on both config object the SessionReplay object. I wonder if we should make a distinction of these properties, i.e. things in `config` are static and would require teardown/re-init if update is needed. Then the dynamic properties can be exposed only on the SessionReplay object.

Reply:
When you call `setSessionId` it changes the session id on `this.config`, there aren't two representations of `sessionId`. See https://github.com/amplitude/Amplitude-TypeScript/blob/78ad90a58053d4971bf54231bcf96b42ff90b6b1/packages/session-replay-browser/src/session-replay.ts#L84-L84

There are three properties on config that can be updated outside of init, `deviceId`, `optOut`, and `sessionId`. We currently provide only an interface on the standalone SDK for changing `sessionId` mid-recording, but `deviceId` and `optOut` can be changed when used as a plugin, if `browserSDK.setDeviceID` or `browserSDK.setOptOut` are called.   I'm definitely open to adding an interface to the standalone sdk to support changing these properties directly as well, we can make that change in a later update I think if we decide to, since it'd additive.

The mutability of some properties on config and immutability of others matches the behavior of the Browser SDK so I'd like to overall keep that consistent!